colima: 0.4.2 -> 0.4.4

tricktron opened the pull request:
###### Description of changes
Edit: ~~The colima derivation needs direct access to the lima binary When the pr https://github.com/NixOS/nixpkgs/pull/179521 is merged,~~ then this pr fixes the no internet in vm issue from here: https://github.com/abiosoft/colima/issues/344.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

tricktron committed:
colima: use updated Makefile in build and install phases

tricktron committed:
colima: use lima-unwrapped

tricktron committed:
colima: fix dynamic version

06kellyjac commented on pkgs/applications/virtualization/colima/default.nix:
```diff
@@ -2,8 +2,12 @@
-, lima
+, lima-unwrapped
 , makeWrapper
+, git
+, perl
+, qemu
+, openssl
```
```suggestion
```

`git` can be removed once reverting the `postFetch` removal

---

`perl` doesn't seem used at all

---

`qemu` can be removed if grabbing the `binPath` from lima e.g.

`lima.pasthru.binpath` & `lima.pasthru.unwrapped`

also means colima could support `extraLimaPackages` and do `ourlima = lima.override { extraPackages = extraLimaPackages; }` and get `binpath` from `ourlima.pasthru.binpath`
or people can pass an already overriden `lima` in

---

`openssl` seems to just be for generating a shasum (as part of the `make build`) which we don't ever use, this can be removed once the `Makefile` is no-longer used in the `buildPhase`

06kellyjac reviewed (changes requested):
Some of these comments will require changes in #179521 to match

06kellyjac commented on pkgs/applications/virtualization/colima/default.nix:
```diff
@@ -13,31 +17,39 @@ buildGo118Module rec {
   postInstall = ''
     wrapProgram $out/bin/colima \
-      --prefix PATH : ${lib.makeBinPath [ lima ]}
+      --prefix PATH : ${lib.makeBinPath [ lima-unwrapped qemu ]}
```
You'd probably still have the two separate derivations, I'm just saying the wrapped derivation could expose some more stuff. I'll try make some changes and give an example

06kellyjac replied:
ok here we go:

colima using `lima` but still has `qemu`

```
λ cat ./result/bin/colima
#! /nix/store/j9rc786ylq8cid4zgcn6idknilkbd4ax-bash-5.1-p16/bin/bash -e
PATH=${PATH:+':'$PATH':'}
PATH=${PATH/':''/nix/store/sa8sd1isamxybk6f55r2k10f2pmpi5pi-qemu-7.0.0/bin'':'/':'}
PATH='/nix/store/sa8sd1isamxybk6f55r2k10f2pmpi5pi-qemu-7.0.0/bin'$PATH
PATH=${PATH#':'}
PATH=${PATH%':'}
export PATH
PATH=${PATH:+':'$PATH':'}
PATH=${PATH/':''/nix/store/90104vaymsdjy62gvxwdr1pz656pcy79-lima-0.11.1/bin'':'/':'}
PATH='/nix/store/90104vaymsdjy62gvxwdr1pz656pcy79-lima-0.11.1/bin'$PATH
PATH=${PATH#':'}
PATH=${PATH%':'}
export PATH
exec -a "$0" "/nix/store/xav70i18613vg4v4c9lnw7p8a77gm2j1-colima-0.4.2/bin/.colima-wrapped"  "$@"

λ ls -al /nix/store/90104vaymsdjy62gvxwdr1pz656pcy79-lima-0.11.1/bin/
dr-xr-xr-x root root 4.0 KB Thu Jan  1 01:00:01 1970  .
dr-xr-xr-x root root 4.0 KB Thu Jan  1 01:00:01 1970  ..
.r-xr-xr-x root root 561 B  Thu Jan  1 01:00:01 1970  docker.lima
.r-xr-xr-x root root 950 B  Thu Jan  1 01:00:01 1970  lima
.r-xr-xr-x root root  12 MB Thu Jan  1 01:00:01 1970  limactl
.r-xr-xr-x root root  41 B  Thu Jan  1 01:00:01 1970  nerdctl.lima
.r-xr-xr-x root root 576 B  Thu Jan  1 01:00:01 1970  podman.lima
```

rather than `lima-wrapper`


```
λ nix-build -A lima
this derivation will be built:
  /nix/store/ajlgddwjcmn5hlgmn6zdypj8mvwc47p9-lima-wrapper-0.11.1.drv
building '/nix/store/ajlgddwjcmn5hlgmn6zdypj8mvwc47p9-lima-wrapper-0.11.1.drv'...
/nix/store/n637krbwz0y4qdjflvzlx26hxnfyv4w5-lima-wrapper-0.11.1
λ ls -al ./result/bin
dr-xr-xr-x root root 4.0 KB Thu Jan  1 01:00:01 1970  .
dr-xr-xr-x root root 4.0 KB Thu Jan  1 01:00:01 1970  ..
.r-xr-xr-x root root 381 B  Thu Jan  1 01:00:01 1970  limactl
λ ls -al ./result
lrwxrwxrwx jk users 63 B Wed Jun 29 15:28:08 2022  ./result ⇒ /nix/store/n637krbwz0y4qdjflvzlx26hxnfyv4w5-lima-wrapper-0.11.1
```

<details>
  <summary>patch</summary>

```patch
diff --git a/pkgs/applications/virtualization/colima/default.nix b/pkgs/applications/virtualization/colima/default.nix
index 4b756bb720a..a112c02e19d 100644
--- a/pkgs/applications/virtualization/colima/default.nix
+++ b/pkgs/applications/virtualization/colima/default.nix
@@ -1,12 +1,12 @@
 { lib
-, buildGo118Module
+, buildGoModule
 , fetchFromGitHub
 , installShellFiles
 , lima
 , makeWrapper
 }:
 
-buildGo118Module rec {
+buildGoModule rec {
   pname = "colima";
   version = "0.4.2";
 
@@ -23,21 +23,27 @@ buildGo118Module rec {
       rm -rf $out/.git
     '';
   };
+  vendorSha256 = "sha256-91Ex3RPWxOHyZcR3Bo+bRdDAFw2mEGiC/uNKjdX2kuw=";
 
   nativeBuildInputs = [ installShellFiles makeWrapper ];
 
-  vendorSha256 = "sha256-91Ex3RPWxOHyZcR3Bo+bRdDAFw2mEGiC/uNKjdX2kuw=";
+  ldflags = [
+    "-s"
+    "-w"
+    "-X github.com/abiosoft/colima/config.appVersion=${version}"
+  ];
 
-  doCheck = false;
 
-  preConfigure = ''
-    ldflags="-X github.com/abiosoft/colima/config.appVersion=${version}
-              -X github.com/abiosoft/colima/config.revision=$(cat .git-revision)"
+  preBuild = ''
+    ldflags+=" -X github.com/abiosoft/colima/config.revision=$(cat .git-revision)"
   '';
 
+  doCheck = false;
+
+  # colima requires an unwrapped lima but we still want the packages lima was wrapped in to be available too
   postInstall = ''
     wrapProgram $out/bin/colima \
-      --prefix PATH : ${lib.makeBinPath [ lima ]}
+      --prefix PATH : ${lib.makeBinPath ([ lima.passthru.lima-unwrapped ] ++ lima.passthru.wrapperPackages)}
 
     installShellCompletion --cmd colima \
       --bash <($out/bin/colima completion bash) \
diff --git a/pkgs/applications/virtualization/lima/default.nix b/pkgs/applications/virtualization/lima/default.nix
index c6b9b4149fe..c9efdb98fab 100644
--- a/pkgs/applications/virtualization/lima/default.nix
+++ b/pkgs/applications/virtualization/lima/default.nix
@@ -2,8 +2,6 @@
 , buildGoModule
 , fetchFromGitHub
 , installShellFiles
-, qemu
-, makeWrapper
 }:
 
 buildGoModule rec {
@@ -19,7 +17,7 @@ buildGoModule rec {
 
   vendorSha256 = "sha256-wvb592mmxOeRSX8Rxx2m7tZ2S8wQTcQkL3b6z0F6OEQ=";
 
-  nativeBuildInputs = [ makeWrapper installShellFiles ];
+  nativeBuildInputs = [ installShellFiles ];
 
   # clean fails with read only vendor dir
   postPatch = ''
@@ -32,12 +30,14 @@ buildGoModule rec {
     runHook postBuild
   '';
 
+  # functions required for the default checkPhase are added in the buildPhase
+  # which we overwrite
+  doCheck = false;
+
   installPhase = ''
     runHook preInstall
     mkdir -p $out
     cp -r _output/* $out
-    wrapProgram $out/bin/limactl \
-      --prefix PATH : ${lib.makeBinPath [ qemu ]}
     installShellCompletion --cmd limactl \
       --bash <($out/bin/limactl completion bash) \
       --fish <($out/bin/limactl completion fish) \
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index d6f1798ad64..df049c57526 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -34458,9 +34458,13 @@ with pkgs;
 
   idsk = callPackage ../tools/filesystems/idsk { stdenv = gcc10StdenvCompat; };
 
-  colima = callPackage ../applications/virtualization/colima { };
+  colima = callPackage ../applications/virtualization/colima {
+    buildGoModule = buildGo118Module;
+  };
+
+  lima = callPackage ../applications/virtualization/lima/wrapper.nix { };
 
-  lima = callPackage ../applications/virtualization/lima { };
+  lima-unwrapped = callPackage ../applications/virtualization/lima { };
 
   logtop = callPackage ../tools/misc/logtop { };
 
```

</details>

it can be made robust so that it can handle being passed an unwrapped lima like this:

```nix
{ pkgs ? import ./. { } }:

pkgs.colima.override { lima = pkgs.lima-unwrapped; }
```

if changed like this:

```
let
  isLimaWrapped = builtins.hasAttr "wrapperPackages" lima.passthru;
  lima-unwrapped = if isLimaWrapped then lima.passthru.lima-unwrapped else lima;
  # colima requires an unwrapped lima but we still want the packages lima was
  # wrapped in (if it was a wrapped lima) to be available too
  packages = [
    lima-unwrapped
  ] ++ lib.optionals (isLimaWrapped) lima.passthru.wrapperPackages;
in
```

&&

```
postInstall = ''
    wrapProgram $out/bin/colima \
      --prefix PATH : ${lib.makeBinPath packages}
# ...
```

I'll push the fully robust one to a branch and link it since that might be easier to scroll through than the patch file output

tricktron replied:
@06kellyjac This is not needed anymore since we fixed it upstream😃.

tricktron committed:
colima: 0.4.2 -> 0.4.3

tricktron committed:
colima: delete .git folder in postfetch when `leaveDotGit = true`

Otherwise it may lead to non-deterministic behaviour.

Co-authored-by: j-k <dev@j-k.io>

aaschmid commented on pkgs/applications/virtualization/colima/default.nix:
```diff
@@ -2,8 +2,12 @@
 , buildGo118Module
```
I changed this some time ago because didn't know better, however your hint is reasonable :-)

tricktron commented:
I'll apply the rest of the suggestions tomorrow👍🏼

Atemu commented on pkgs/applications/virtualization/colima/default.nix:
```diff
@@ -13,31 +17,39 @@ buildGo118Module rec {
     leaveDotGit = true;
-    postFetch = ''
-      git -C $out rev-parse HEAD > $out/.git-revision
-      rm -rf $out/.git
```
WDYM? .git is still being deleted just like before.

06kellyjac replied:
this comment was made several force pushes ago

Atemu replied:
Oh, odd that GitHub didn't label it as outdated.

tricktron committed:
colima: do not override default buildGoModule phases

The colima Makefile does more or less the same as the `buildGoModule`
implementation. Instead of overriding it and using the Makefile
directly, we reproduce the behaviour desired behaviour by customising
env variables, e.g. `ldflags` and `subPackages`.

Co-authored-by: j-k <dev@j-k.io>

tricktron commented on pkgs/applications/virtualization/colima/default.nix:
```diff
@@ -49,6 +61,6 @@ buildGo118Module rec {
     description = "Container runtimes on MacOS with minimal setup";
```
@abiosoft
Is colima also targeted for linux/does it work on linux?

tricktron committed:
colima: use `buildGoModule = buildGo118Module;`

Co-authored-by: j-k <dev@j-k.io>

tricktron committed:
colima: update meta.description

Co-authored-by: j-k <dev@j-k.io>

tricktron committed:
colima: set `CGO_ENABLED = 1` to make VPN connections work

See https://github.com/abiosoft/colima/issues/358

tricktron commented on pkgs/applications/virtualization/colima/default.nix:
```diff
@@ -2,8 +2,12 @@
 , fetchFromGitHub
 , installShellFiles
-, lima
+, lima-unwrapped
```
@06kellyjac This is not needed anymore since we fixed it upstream:smiley:.

Atemu commented on pkgs/applications/virtualization/colima/default.nix:
```diff
@@ -1,43 +1,50 @@
+  ldflags = [
+    "-s"
+    "-w"
+    "-X github.com/abiosoft/colima/config.appVersion=${version}"
+  ];
 
-  preConfigure = ''
-    ldflags="-X github.com/abiosoft/colima/config.appVersion=${version}
-              -X github.com/abiosoft/colima/config.revision=$(cat .git-revision)"
+  preBuild = ''
+    ldflags+=" -X github.com/abiosoft/colima/config.revision=$(cat .git-revision)"
```
I think it's better to define ldflags in one place rather than two like I did it before.

tricktron replied:
@Atemu, works for me.

tricktron committed:
colima: 0.4.3 -> 0.4.4

tricktron committed:
colima: define all ldflags in preConfigure

Co-authored-by: Atemu <Atemu@users.noreply.github.com>

tricktron committed:
colima: add version test

tricktron commented:
@ofborg build colima.passthru.tests.version

tricktron commented:
@ofborg build colima

tricktron commented:
@06kellyjac @Atemu Ready for second round of review if you want.

What is the default value for meta.platforms? Why is ofborg not building it on aarch64-darwin anymore? I remember that this worked some time ago.

tricktron committed:
colima: add meta.mainProgram and meta.platforms

tricktron commented:
@ofborg build colima.passthru.tests.version

tricktron commented:
@ofborg build colima.passthru.tests.version

Atemu commented on pkgs/applications/virtualization/colima/default.nix:
```diff
@@ -1,54 +1,66 @@
     license = licenses.mit;
-    maintainers = with maintainers; [ aaschmid ];
+    mainProgram = "colima";
+    platforms = [ "aarch64-linux" "aarch64-darwin" "x86_64-linux" "x86_64-darwin" ];
```
```suggestion
    platforms = with platforms; linux ++ darwin;
```

or something like that

tricktron replied:
@Atemu 
Yes, I usually did that but from other reviews I got the feedback that I should use the above because
`platforms.linux ++ platforms.darwin = [ "aarch64-linux" "armv5tel-linux" "armv6l-linux" "armv7a-linux" "armv7l-linux" "i686-linux" "m68k-linux" "mipsel-linux" "mips64el-linux" "powerpc64-linux" "powerpc64le-linux" "riscv32-linux" "riscv64-linux" "s390-linux" "s390x-linux" "x86_64-linux" "x86_64-darwin" "i686-darwin" "aarch64-darwin" "armv7a-darwin" ]` results in a long list and I don't know if every platform is supported.

Guess it makes more sense to add supported platforms in the future than removing unsupported platforms due to bugs. What do you think? Is there a general consensus?

SuperSandro2000 replied:
and linux + darwin is the default so we can drop it.

> but from other reviews I got the feedback that I should use the above because

thats not really how we do things.

> results in a long list and I don't know if every platform is supported.

I can guarantee you that it isn't but thats okay. We don't want to touch the meta sections of thousands of packages if a new architecture gets great support.

Generally this is mostly about if a package is supported on linux or darwin and x86_64 or aarch64. All the arm variants, m68k, mipsel or s390 don't really matter especially that we don't have hydra support for them.

tricktron replied:
@SuperSandro2000 Ok, I removed the meta.platforms attribute again:+1:.

Atemu replied:
@tricktron FYI, you didn't.

tricktron replied:
@Atemu Did you see my latest commit 10 hours ago https://github.com/NixOS/nixpkgs/commit/62b5ad1363dc5a70edf1a630ca0939ab2b0c6a2f?

06kellyjac commented on pkgs/applications/virtualization/colima/default.nix:
```diff
@@ -1,54 +1,66 @@
+, qemu
+, testers
+, colima
+, runCommand
```
Is this used?

abiosoft replied:
looks more like forgotten debug lines

tricktron replied:
@06kellyjac
runCommand slipped through, thanks for catching that.

- qemu is used by the wrapper
- testers and colima (the package itself) are used in passthru for testing. Adopted from here https://nixos.org/manual/nixpkgs/stable/#chap-testers

06kellyjac replied:
yeah the comment was just for the runcommand line, not sure why it grabbed the whole section

tricktron committed:
colima: remove unused runCommand input

Co-authored-by: j-k <dev@j-k.io>

SuperSandro2000 commented on pkgs/applications/virtualization/colima/default.nix:
```diff
@@ -1,54 +1,65 @@
   preConfigure = ''
-    ldflags="-X github.com/abiosoft/colima/config.appVersion=${version}
-              -X github.com/abiosoft/colima/config.revision=$(cat .git-revision)"
+    ldflags="-s -w -X github.com/abiosoft/colima/config.appVersion=${version} \
+    -X github.com/abiosoft/colima/config.revision=$(cat .git-revision)"
```
Please set everything static in ldflags attribute.

Atemu replied:
Not possible because the git revision needs to be read from a file during the build.

06kellyjac replied:
I prefer static where possible and the extra stuff in prebuild but thats just me 🤷

Atemu replied:
Problem is that we'd need to modify the statically declared env variable in preBuild which I think is even less clean than just setting it in preBuild to begin with.

tricktron replied:
@SuperSandro2000 @06kellyjac @Atemu
How shall we proceed here? I am ok with both approaches:
- static ld flags + changing it dynamically in prebuild
- all in configurePhase like right now

06kellyjac replied:
It doesnt matter too much, whichever you'd prefer since you'll be maintaining it :)

tricktron replied:
Ok, then I'll leave the ldflags as is in the configurePhase.

SuperSandro2000 replied:
> whichever you'd prefer since you'll be maintaining it :)

well yes but that also does not mean you can do whatever.

SuperSandro2000 commented on pkgs/applications/virtualization/colima/default.nix:
```diff
@@ -1,54 +1,65 @@
     homepage = "https://github.com/abiosoft/colima";
     license = licenses.mit;
-    maintainers = with maintainers; [ aaschmid ];
+    mainProgram = "colima";
```
```suggestion
```

tricktron committed:
colima: cleanup meta

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

tricktron committed:
colima: drop meta.platforms

because we use the default value of `platforms.darwin ++
platforms.linux`

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>
Co-authored-by: Atemu <Atemu@users.noreply.github.com>

tricktron commented:
@ofborg build colima colima.passthru.tests.version

tricktron commented:
@SuperSandro2000 @06kellyjac @Atemu
In my opinion this is ready to merge.

SuperSandro2000 commented on pkgs/applications/virtualization/colima/default.nix:
```diff
@@ -1,54 +1,63 @@
   nativeBuildInputs = [ installShellFiles makeWrapper ];
 
-  vendorSha256 = "sha256-91Ex3RPWxOHyZcR3Bo+bRdDAFw2mEGiC/uNKjdX2kuw=";
+  vendorSha256 = "jDzDwK7qA9lKP8CfkKzfooPDrHuHI4OpiLXmX9vOpOg=";
```
```suggestion
  vendorSha256 = "sha256-jDzDwK7qA9lKP8CfkKzfooPDrHuHI4OpiLXmX9vOpOg=";
```

SuperSandro2000 committed:
Update pkgs/applications/virtualization/colima/default.nix
